### ISSUE TITLE
[Meta] Observability TSDB packages migration

### ISSUE THREAD

**Issue by rameshelastic** [elastic/integrations#5233]
# Migration summary

Important resources:

- Spreadsheet listing [package prioritisation ](https://docs.google.com/spreadsheets/d/1bw2De7q1-dnSxCaKFvrPSTSDEGbYHLXbMuaIT8KyOHY/edit?usp=sharing)(private link)

- Packages are being migrated dataset wise, following the generic guideline [here](https://github.com/elastic/integrations/blob/main/docs/developer_tsdb_migration_guidelines.md). The scope of the migration is to cover both addition of [dimensions](https://github.com/elastic/integrations/blob/main/docs/generic_guidelines.md#specify-dimensions) and [metric types](https://www.elastic.co/guide/en/elasticsearch/reference/current/tsds.html#time-series-metric).

- Mandatory ECS fields: https://github.com/elastic/integrations/issues/5193#issuecomment-1536090359

- Testing the migration: https://github.com/elastic/TSDB-migration-test-kit

##  Packages ongoing

- [ ] system https://github.com/elastic/integrations/issues/5848
- [ ] Elasticsearch: https://github.com/elastic/integrations/issues/6618
  - 7 / 10 migrated. The 3 missing are blocked.
- [ ] Elastic agent: https://github.com/elastic/integrations/issues/6938
- [ ] Kibana: https://github.com/elastic/integrations/issues/7024
- [ ] Azure: https://github.com/elastic/integrations/issues/7140
- [ ] Windows: https://github.com/elastic/integrations/issues/6993 
- [ ] Couchbase: https://github.com/elastic/integrations/issues/7378
- [ ] GCP: https://github.com/elastic/integrations/issues/7555
- [ ] Zookeeper https://github.com/elastic/integrations/issues/7752
- [ ] Hashicorp Vault https://github.com/elastic/integrations/issues/8128
- [ ] HA Proxy: https://github.com/elastic/integrations/issues/7138
- [ ] NATS : https://github.com/elastic/integrations/issues/7745
- [ ] STAN : https://github.com/elastic/integrations/issues/7859
- [ ] ETCD : https://github.com/elastic/integrations/issues/8327
- [ ] Containerd : https://github.com/elastic/integrations/issues/8613

##  Completed packages
- [x] Apache Spark https://github.com/elastic/integrations/issues/7786
- [x] Oracle WebLogic: https://github.com/elastic/integrations/issues/7512
- [x] Traefik https://github.com/elastic/integrations/issues/7766
- [x] Oracle https://github.com/elastic/integrations/issues/5310
- [x] InfluxDB https://github.com/elastic/integrations/issues/5283
- [x] Ngnix https://github.com/elastic/integrations/issues/5282
- [x] CockroachDB https://github.com/elastic/integrations/issues/5410
    - metric_type pending on https://github.com/elastic/kibana/issues/155004
- [x] RabbitMQ https://github.com/elastic/integrations/issues/5279
- [x] MSSQL https://github.com/elastic/integrations/issues/5278
- [x] Redis https://github.com/elastic/integrations/issues/5698
- [x] Airflow https://github.com/elastic/integrations/issues/5275
    - metric_type pending on https://github.com/elastic/kibana/issues/155004
- [x] CouchDB https://github.com/elastic/integrations/issues/5854
- [x] Kafka https://github.com/elastic/integrations/issues/5905
- [x] Kubernetes https://github.com/elastic/integrations/issues/4618
    - Migration is complete and validation on-going: https://github.com/elastic/obs-infraobs-team/issues/1007
    - [ ] Agent.id is not a dimension in this migration. Pending?
- [x] IIS https://github.com/elastic/integrations/issues/6267
- [x] Vsphere https://github.com/elastic/integrations/issues/6266
- [x] Docker: https://github.com/elastic/integrations/issues/6393
- [x] Apache https://github.com/elastic/integrations/issues/6075
- [x] ActiveMQ https://github.com/elastic/integrations/issues/6473
- [x] IBM MQ https://github.com/elastic/integrations/issues/6464
- [x] MongoDB https://github.com/elastic/integrations/issues/5281
- [x] AWS Fargate: https://github.com/elastic/integrations/issues/6732
- [x] Istio: https://github.com/elastic/integrations/issues/6567
- [x] AWS: for all metrics datastreams https://github.com/elastic/integrations/issues/6293
- [x] Prometheus https://github.com/elastic/integrations/issues/5800
- [x] Postgres https://github.com/elastic/integrations/issues/5429
- [x] MySQL https://github.com/elastic/integrations/issues/5277
- [x] Memcached : https://github.com/elastic/integrations/issues/6925

##  Blocked packages
- [ ] GCP Firestore https://github.com/elastic/integrations/issues/8620

# User reported issues

- https://github.com/elastic/integrations/issues/6554
- https://github.com/elastic/sdh-beats/issues/3536 TSDB got enabled on logs-* data stream
- https://github.com/elastic/integrations/issues/6804
- https://github.com/elastic/integrations/issues/6996 
- https://github.com/elastic/sdh-beats/issues/3780
  - Corresponding beats issue https://github.com/elastic/beats/issues/36547  agent re-try logic causing 409 error

# Internal o11y issues found
- **Blocker**: Issue impacting underlying TSDB functionality for multiple packages. It would require a fix/back-port for a release or a committed timeline.
- **Critical**: Issue impacting multiple packages, sometime causing the gating of the migration activity for a given package. However, release gating may not be done.

| Source                                      | Type     | Summary                                                                 | Severity            |  Status |   Next step |
| :--------------------               | :---     | :------------------------------------------------- | :--------------- | :-------------- |-----------------|
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/99176) | Documentation | Document how to reindex a TSDB datastream | Critical | Open |  Need documentation for TSDB reindexing | 
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/98384) | Enhancement | Add an option to treat non-metric fields as a dimension by default | Critical | Open |  Currently WIP. | 
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/98463) 	| Bug 	|  Old timestamp documents being dropped | Critical 	| Closed |  Elasticsearch [provided ability](https://github.com/elastic/elasticsearch/pull/98518) to configure look_back_time. Long term issue captured separately. 
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/99343) 	| Bug 	|  Old timestamp documents being dropped | Medium 	| Open |  Long term issue tracking delayed document arrival issue. 
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/99123) 	| Enhancement 	|  Adding metric names to TSID | Medium 	| Open |  Long term issue. 
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/98384) 	| Enhancement 	|  Add an option to treat non-metric fields as a dimension by default |  Critical | Open |  Long term issue, more applicable for OTel documents ingestion. 
| [elasticsearch](https://github.com/elastic/kibana/issues/160040) 	| Bug 	| Improve error handing via documentation for custom dashboard being broken  | Critical 	| Validation WIP | 8.10 elasticsearch updated to support all counter aggregations. Back ported to 8.92. being tested [here](https://github.com/elastic/kibana/issues/160040#issuecomment-1700813686).
| [elasticsearch](https://github.com/elastic/kibana/issues/161369#issuecomment-1625344541) 	| Bug 	| Certain type of counter aggregations are not supported	| Critical 	| Open 	| There are certain counter aggr. currently not supported. detailed provided, used by k8/aws package  |
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/97394) 	| Bug 	| `scaled_float` fields are not searchable.  Related [issue](https://github.com/elastic/integrations/issues/6804)	| Critical 	| Closed | Issue with released system package, 8.10 fixed elasticsearch team |
| [integrations](https://github.com/elastic/integrations/issues/6568)           | Bug         | GCP metrics are not grouped by labels | Critical |  Open | Bug fix expected to resume |
| [kibana](https://github.com/elastic/kibana/issues/157839)           | Bug         | Pie visualisation issue for non-tsdb and tsdb data combined. | Medium |  Open | Discussions on-going and we are sharing details on the field mapping. |
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/96552)           | Bug         | Numbers that are mapped as keyword are not validated correctly at routing time | Medium |  Open | Discussion on-going. |
| [elasticsearch](https://github.com/elastic/kibana/issues/160040)           | Critical         | Custom dashboard failing aggregation | Critical |  Validation TBD | 8.10 elasticsearch updated to support all counter aggregations. Earlier [thread](https://github.com/elastic/elasticsearch/issues/96169). Duplicate older issue. |
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/98157)           | Doc         | Facing issue while reindexing a TSDB enabled data stream | Medium |  Open |  Captured separately, only document change needed. https://github.com/elastic/elasticsearch/issues/99176 |
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/98171)           | Enhancement         | Support GCP delta metrics | Medium |  Open | Map to gauge for now. |
| [elasticsearch](https://github.com/elastic/integrations/issues/6604)           | Bug         | Issue enabling TSDB on nested fields | Critical |  Open | Need confirmation if ES issue |
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/96169)           | Bug         | Provide a mechanism for users to go back to old working custom dashboards | Critical |  Closed | Duplicate of [issue](https://github.com/elastic/elasticsearch/issues/96170). [Testcase](https://github.com/elastic/elasticsearch/issues/96170#issuecomment-1560944214) to reproduce. Shard errors coming in the TSDB mode. Should there be better error message? |
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/96254)           | Bug         | Text fields in data stream has error | Medium |  Open | Exploring the mapping suggestion for stored field |
| [elastic-package](https://github.com/elastic/elastic-package/issues/1111)           | Enhancement         | Documentation annotate fields that are dimensions and metric types | Medium |  Open | Low priority for now. |
| [Fleet](https://github.com/elastic/ingest-docs/issues/151)           | Enhancement         | Need manual dev tools based steps for tsdb toggle on/of | Critical |  Closed | Manual TSDB [disable](https://github.com/elastic/ingest-docs/issues/151#issuecomment-1547447587) has been confirmed to work. Need official docs to be released, so that manual TSDB disable can be enabled. |
| [kibana](https://github.com/elastic/kibana/issues/157239)           | Bug         | Discover fields UI does not change icons counter/gauge field after package upgrade to TSDB | Critical |  Open | Not TSDB related. User need to open new browser windows after package upgrade as an work-around. |
| [kibana](https://github.com/elastic/kibana/issues/155510)           | Enhancement         | Counter filed show- Analysis not available in discover | Medium |  Closed | Kibana team is working on replacing this current message with a better message suitable for counter fields. |
| [elastic-package](https://github.com/elastic/elastic-package/issues/1253)         | Bug         | Support fields validation when synthetic source is used | Critical |  Open | Not many packages are impacted |
| [ECS](https://github.com/elastic/integrations/issues/5193)	| Enhancement 	| ECS centralisation of common dimension fields 	| Critical 	| Open 	| Definition of common dimension fields(elastic/integrations#5193). ECS stage 0 [merged](https://github.com/elastic/ecs/pull/2172). Recent updates done.|
| [k8 package](https://github.com/elastic/kibana/issues/155587) 	| Enhancement 	| Certain aggregation functions for counters in k8 package needs to be updated 	| Critical 	| Open 	| Since we have earlier used sum() aggr functions for some of the counter type fields, in k8 package dashboard. These need to be updated, as sum is not supported for counters in TSDB. |
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/93539) 	| Bug 	| Certain aggregation functions not working for counters 	| Critical 	| Partially fixed 	| The gating ones fixed. Still some functions(median/avg) not supported. Need to close with ES team what is the overall plan for the counter aggregation functions.|
| [elasticsearch](https://github.com/elastic/elastic-package/issues/1174) 	| Bug 	|Dimension limit override settings blocking the build 	|Medium 	|Closed 	|Only for packages with kibana.version <8.6, No issue if package version of made => 8.6. This impacts InfluxDB and may be few other prometheus type packages. Closing this as not applicable anymore.|
| [Kibana](https://github.com/elastic/kibana/issues/149180) 	|Enhancement 	|Need UI indicators for dimension fields when TSDB is enabled 	|Medium 	|Open 	|Kibana team to plan.|
| [Kibana](https://github.com/elastic/kibana/issues/150549) 	|Enhancement 	|Display the current write index when timeseries mode is enabled 	|Medium 	|Open 	|Kibana team to plan.|
| [elasticsearch](https://docs.google.com/document/d/1KO52R0TKcGfPtlVeKYHZfT11VULrG2At9_jDTKS-mP0/edit) | Bug |Histograms in TSDS mode don’t include zero counts 	| Medium 	|Closed 	|The conclusion is that no tool changes needed and the behaviour should not impact user in way. Only documentation to be updated.|
| [kibana](https://github.com/elastic/kibana/issues/152467) 	|Bug 	|Counter fields don’t show in Visualisation. Duplicate [issue](https://github.com/elastic/elasticsearch/issues/95031) 	|Blocker 	|Closed 	| 8.7 the issue exist and kibana team is targeting 8.8 fix. Nginx and MSSQL & k8 packages verified in 8.8 snapshot to show counter fields in visualisation.|
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/94380) 	| Bug 	| IP is not supported as dimension field 	| Medium 	| Closed 	| Verified. 8.8.0-SNAPSHOT|
| [Fleet](https://github.com/elastic/kibana/issues/153220) 	|Bug 	| Fields having type byte, short, integer, long, unsigned_long are not mapped correctly as dimension fields 	| Medium 	| Closed 	| Verified. 8.8.0-SNAPSHOT|
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/94113) 	| Bug 	| Enablement of TSDB fails for package with fields having flattened type 	| Critical 	| Fixed 	| [Verified](https://github.com/elastic/elasticsearch/issues/94113#issuecomment-1514134532) to work 8.8 snapshot. Now dependant packages can be committed with 8.8 stack version.|
| [elasticsearch](https://github.com/elastic/elasticsearch/issues/96163) 	| Bug 	| Immediately upgrading a downgraded tsdb data stream fails 	| Critical 	| Not an issue 	| 4 hours gap needed. Not an issue for now |
| [Fleet](https://github.com/elastic/kibana/issues/157345) 	| Bug 	| Does not allow package upgrade to disable TSDB 	|Blocker 	|Closed 	|This has been fixed in latest snapshot. Separate PR for fixing rollover has been verified to work|
| [Fleet](https://github.com/elastic/kibana/issues/160403)           | Bug         | Mapping does not reflect the fields definition in package | Critical |  Closed | Double field definition, closed.|
| [Fleet](https://github.com/elastic/kibana/issues/155004)           | Enhancement         | Need metric_type mapping for dynamic templates | Critical |  Closed | Closed This impacts migration of multiple Prometheus style packages. Fixed from 8.9 |

**Comment by agithomas** [comment]
Key question : Should timeseries enablement be done on datastream type of type `logs` ? 


If the answer is `yes`, please go through the below mentioned situation. If the answer is 'No',  please go to **Case B**

**Case A : TSDB enablement is needed on logs datastream** 
When it comes to datastream of type `logs`, we may come across 3 types of logs

1. access logs
2. error logs
3. audit trial logs

Sharing an observation (can become a potential problem as well) and a usecase issue here

**Observation :** 

Logs of various type and structure may fall into access log, error log .  Certain integrations such as `nginx` capture only the ip address from these log entries. If timeseries is enabled , it would end up in data loss scenario as there do not exist enough `DIMENSION` fields to isolate a log entry from another from being overwritten. Data loss is highly probable.


**Usecase**

Logs such as audit trails are not centered around metrics, have keyword type fields used more often. Are there any benefit in TSDB enablement in an audit-trail usecase scenario ? 

**Case B : TSDB Enablement is NOT needed on logs datastream**

It may be best not to show TSDB enablement in Kibana UI when datastream type is `logs`. 

Considering the challenges mentioned under Case A, i do not think it is beneficial to target datastream that are of type `logs`.  @ruflin , @lalit-satapathy , please share your thoughts on this.

**Comment by ruflin** [comment]
> Question : Should timeseries enablement be done on datastream type of type logs ?

No. I recommend to ignore everything around logs right now.

> In Kibana UI, there must not be a selection when a datastream is having a type value log.

Agree. The flags will be reworked in the upcoming releases ( @kpollich )

> If there exist no candidates for metric_type, the specific datastream does not qualify for TSDB migration.

This is an interesting point. Do you expect some `metrics-*-*` data streams to have no `metric_type` fields?

**Comment by agithomas** [comment]
> This is an interesting point. Do you expect some `metrics-*-*` data streams to have no `metric_type` fields?

I am suggesting this as a general rule to follow irrespective log type or metric type. 

This is because, there may exist a few datastreams having type based on http_json ( type : log )  having metric_type value gauge. 

Since we don't have all the `metrics-*-*` datastreams annotated with `metric_type`, i am unable to extract the statistics .  Can there be a metrics based datastream having no metric_type ?  It is highly unlikely, but certainly not impossible.

**Comment by ruflin** [comment]
> This is because, there may exist a few datastreams having type based on http_json ( type : log ) having metric_type value gauge.

Should these be `type: metrics`?

**Comment by agithomas** [comment]
> Should these be `type: metrics`?

@lalit-satapathy , i believe, there was a discussion on this topic , much before TSDB migration started, regarding this topic. Was any decisions made?

**Comment by lalit-satapathy** [comment]
The _logs_ data stream having _metrics_ data is a **small set** and are not in the packages being considered for migration currently. These handful integrations are built with http json [input](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-httpjson.html). 

There is an [issue](https://github.com/elastic/elastic-agent/issues/2208) tracking why, we are not able to overwrite _type_ for these packages to _metrics_.

**Comment by ruflin** [comment]
For now, lets focus only on `metrics-*-*` indices.

**Comment by lalit-satapathy** [comment]
Updated meta [issue](https://github.com/elastic/integrations/issues/5233#issue-1579513326) to track all O11y TSDB packages. Segregating packages to _ongoing/completed/blocked_ and  issues to _elasticsearch/Kibana/Others_.

@tetianakravchenko @constanca-m @mlunadia

**Comment by lalit-satapathy** [comment]
Updated **_Issue Summary_** in [meta](https://github.com/elastic/integrations/issues/5233#issue-1579513326) to: clearly call out **blocker/critical** and also track what is the pending action.  Also, cleaned-up old issues and updated severity as applicable.

**Comment by lalit-satapathy** [comment]
Meta [issue](https://github.com/elastic/integrations/issues/6031) tracking non-cloud ecs changes as per the [recent](https://github.com/elastic/integrations/issues/5193#issuecomment-1536090359) ECS Updates.

**Comment by lalit-satapathy** [comment]
Updated **"Issue Summary"** with latest [issues](https://github.com/elastic/integrations/issues/5233#issue-1579513326), one [blocker](https://github.com/elastic/kibana/issues/157345) marked.

**Comment by mlunadia** [comment]
@nimarezainia @joshdover do you think the 4 issues ( 1 blocker and 3 critical) we have raised for the adoption of TSDB in observability are solvable for 8.8.0? 
- https://github.com/elastic/kibana/issues/157345
- https://github.com/elastic/elastic-package/issues/1253
- https://github.com/elastic/kibana/issues/155004
- https://github.com/elastic/ingest-docs/issues/151

See [table above](https://github.com/elastic/integrations/issues/5233#issue-1579513326) for reference of all issues found and raised. @giladgal who from the fleet team is involved in TSDB?

**Comment by lalit-satapathy** [comment]
Updated "Issue Summary" with latest issues.

**Comment by lalit-satapathy** [comment]
Closing the top level meta issue.
